### PR TITLE
test error message improvements for name inputs. new feature

### PR DESCRIPTION
**Pull request by DanielOsypenko**
Tests cover error message improvements feature https://issues.redhat.com/browse/RHSTOR-4117  for Bucket class name, NamespaceStore name, System storage class name, Object bucket claim name, Blocking pool name, Backing store name.

Tests are UI only.

Known bug fails the test run - https://bugzilla.redhat.com/show_bug.cgi?id=2193109

Multiple tests performed to evaluate performance with given boundary values for the input rules:

OBC

'No more than 253 characters											
'Starts and ends with a lowercase letter or number					
'Only lowercase letters, numbers, non-consecutive periods, or hyphens	
'Cannot be used before												

Backing store

No more than 43 characters											
Starts and ends with a lowercase letter or number					
Only lowercase letters, numbers, non-consecutive periods, or hyphens	
A unique name for the BackingStore within the project				

Bucket class

3-63 characters														
Starts and ends with a lowercase letter or number					
Only lowercase letters, numbers, non-consecutive periods, or hyphens	
Avoid using the form of an IP address								
Cannot be used before												

Namespace store

No more than 43 characters											
Starts and ends with a lowercase letter or number					
Only lowercase letters, numbers, non-consecutive periods, or hyphens	
A unique name for the NamespaceStore within the project				

Blocking pool

No more than 253 characters											
Starts and ends with a lowercase letter or number					
Only lowercase letters, numbers, non-consecutive periods, or hyphens	
Cannot be used before												

StorageSystem name 

No more than 253 characters											
Starts and ends with a lowercase letter or number					
Only lowercase letters, numbers, non-consecutive periods, or hyphens	
Cannot be used before

### CODE REVIEW

File: tests/ui/test_error_improvements.py
Comment (ebenahar) [inline tests/ui/test_error_improvements.py:17]: as discussed, please replace `@ui` with tier3

Comment (DanielOsypenko) [reply tests/ui/test_error_improvements.py:17]: can we have both ui and tier3? or better to remove ui?

File: ocs_ci/ocs/constants.py
Comment (ebondare) [inline ocs_ci/ocs/constants.py:2066]: Do we really need to rename rule1 as rule2, rule5 as rule1 and so on? Wouldn't it be nicer to have some consistency?

Comment (DanielOsypenko) [reply ocs_ci/ocs/constants.py:2066]: all the keys are sorted by numbers "rule1", "rule2", "rule3" for every specific list of rules but not general list of rule.
UI_INPUT_RULES_GENERAL are general rules repeated from one popup to another, thus we use them when required and do not repeat ourself.
we call UI_INPUT_RULES_GENERAL when need to take a rule from common list. This list can not be consistent both for general rules UI_INPUT_RULES_GENERAL and for all the specific case rules such as UI_INPUT_RULES_BUCKET_CLASS, UI_INPUT_RULES_OBJECT_BUCKET_CLAIM or UI_INPUT_RULES_STORAGE_SYSTEM
---------
calling specific rule, such as UI_INPUT_RULES_BUCKET_CLASS or UI_INPUT_RULES_NAMESPACE_STORE or any else we rely on consequent num "rule1", "rule2", "rule3", etc and do not see underlying relation with UI_INPUT_RULES_GENERAL. Basically UI_INPUT_RULES_GENERAL should not be used outside of the constants file

**Comment by DanielOsypenko** [comment]
test run after merge ---> 
https://docs.google.com/document/d/1gDzoI3_BXHM8XzgJ7bdeatz6dVerws7uneAhJYabhqE/edit?usp=sharing

Tests failed due to a bug, all checks passed without unexpected failures.
Tests have bugzilla mark, it will be removed after bug will be fixed.

**Comment by DanielOsypenko** [comment]
minor merge conflict in exceptions file resolved.
Please validate PR
@ebenahar @ebondare @shyRozen @OdedViner @am-agrawa

File: ocs_ci/ocs/ui/mcg_ui.py
Comment (am-agrawa) [inline ocs_ci/ocs/ui/mcg_ui.py:407]: Can we move the import to the top?

File: ocs_ci/ocs/exceptions.py
Comment (am-agrawa) [inline ocs_ci/ocs/exceptions.py:611]:
```suggestion
class IncorrectUIOptionRequested(Exception):
```

File: ocs_ci/ocs/constants.py
Comment (am-agrawa) [inline ocs_ci/ocs/constants.py:2075]: What does `Cannot be used before` means? The meaning isn't very clear.

File: ocs_ci/ocs/constants.py
Comment (am-agrawa) [inline ocs_ci/ocs/constants.py:2083]:
```suggestion
    "rule4": UI_INPUT_RULES_GENERAL["rule3"],
```
Can this be done?

File: ocs_ci/ocs/ui/views.py
Comment (am-agrawa) [inline ocs_ci/ocs/ui/views.py:944]: Can we simply this Xpath?

File: ocs_ci/ocs/constants.py
Comment (DanielOsypenko) [reply ocs_ci/ocs/constants.py:2090]: same as above, this text passed discussions and got approval from management

File: ocs_ci/ocs/ui/mcg_ui.py
Comment (DanielOsypenko) [reply ocs_ci/ocs/ui/mcg_ui.py:407]: Importing it from global creates an import loop, since helpers_ui imports BaseUi

File: ocs_ci/ocs/ui/views.py
Comment (DanielOsypenko) [reply ocs_ci/ocs/ui/views.py:944]: no, this calls axis. It bonds two elements: element with text and it's closest input.
We will use this type of locators more since Ui don't have id's for most of locators and all repeated html nodes created dynamically by JS

File: ocs_ci/ocs/constants.py
Comment (DanielOsypenko) [reply ocs_ci/ocs/constants.py:2083]: I am removing "rule4": "A unique name for the BackingStore within the project from UI_INPUT_RULES_GENERAL, it is actually unique to UI_INPUT_RULES_BACKING_STORE

**Comment by DanielOsypenko** [comment]
passed locally -> http://pastebin.test.redhat.com/1101014 please review @am-agrawa @ebondare @OdedViner @shyRozen

File: tests/ui/test_error_improvements.py
Comment (DanielOsypenko) [reply tests/ui/test_error_improvements.py]: changed the test names

File: ocs_ci/ocs/ui/base_ui.py
Comment (am-agrawa) [inline ocs_ci/ocs/ui/base_ui.py:991]: I am curious to know why is this class in base_ui.py? Is this really a base for all UI tests, I don't think so.

Comment (DanielOsypenko) [reply ocs_ci/ocs/ui/base_ui.py:991]: Same reason why PageNavigator initially is in base_ui.py. CreateResourceForm belongs to multiple resources and the only related parent to it is a PageNavigator

Classically the page classes should be separated by their own files, but python allows to store them all together and some will find it more convenient than to track multiple files across framework.

File: ocs_ci/ocs/exceptions.py
Comment (DanielOsypenko) [reply ocs_ci/ocs/exceptions.py]: > May be this was missed

image

File: ocs_ci/ocs/ui/base_ui.py
Comment (am-agrawa) [reply ocs_ci/ocs/ui/base_ui.py:991]: But why not create it somewhere else and inherit `PageNavigator` they way you are doing it here inside base_ui.py. Wouldn't that be ideal?
Existing `PageNavigator` is most commonly used across various tests, but I don't think that the case with `CreateResourceForm`. It's totally on use case. WYDT?

Comment (DanielOsypenko) [reply ocs_ci/ocs/ui/base_ui.py:991]: @am-agrawa same reason why we use local imports for format_locator - Circular import.
To do so we need to redesign and extract all the classes from base_ui

```
from ocs_ci.ocs.ui.base_ui import PageNavigator
E   ImportError: cannot import name 'PageNavigator' from partially initialized module 'ocs_ci.ocs.ui.base_ui' (most likely due to a circular import) (/Users/danielosypenko/Work/automation_4/ocs-ci/ocs_ci/ocs/ui/base_ui.py)
```
This task will be out of scope of current PR
I wrote myself a ticket https://github.com/red-hat-storage/ocs-ci/issues/7704 I plan complete this task after I finish with feature (I have a close deadline for 2 features)

Comment (DanielOsypenko) [reply ocs_ci/ocs/ui/base_ui.py:991]: @am-agrawa 
adding clarification -
we can not import anything to base_ui.py and it's BaseUI descendants such as PageNavigator, OverviewTab, etc from outside of the file, because these classes themself inherit from BaseUI. Having them all in one file make "import" not needed

Comment (am-agrawa) [reply ocs_ci/ocs/ui/base_ui.py:991]: I am referring to do something similar to `ocs-ci/ocs_ci/ocs/ui/pvc_ui.py`

Comment (DanielOsypenko) [reply ocs_ci/ocs/ui/base_ui.py:991]: it is not similar, because all the classes where I would need to import CreateResourceForm are in the base_ui.py hence it will create circular import. Here is the [article](https://agilno.com/blog/why-cyclic-dependency-errors-occur-a-look-into-the-python-import-mechanism/) about it. More about our case here -> https://github.com/red-hat-storage/ocs-ci/issues/7704 Also, the stuck trace I copied above is from my try to accomplish your request.

Comment (am-agrawa) [reply ocs_ci/ocs/ui/base_ui.py:991]: okay!

**Comment by am-agrawa** [comment]
Rest all looks good to me, will approve once @OdedViner gets a chance to look at it.

**Comment by DanielOsypenko** [comment]
local test run after merge conflict solved -> http://pastebin.test.redhat.com/1101234

File: ocs_ci/ocs/ui/base_ui.py
Comment (OdedViner) [inline ocs_ci/ocs/ui/base_ui.py:854]: Why do you return `obcUI` instance on `navigate_object_bucket_claims_page` method?

**Review by OdedViner — APPROVED** [review]

**Pull request merged**
